storagemarket/integration_test.go

```diff
@@ -286,6 +286,97 @@ func TestMakeDealNonBlocking(t *testing.T) {
 	}, 1*time.Second, 100*time.Millisecond, "actual deal status is %s", storagemarket.DealStates[pd.State])
 }
 
+func TestRestartOnlyProviderDataTransfer(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	h := newHarness(t, ctx, true, noOpDelay, noOpDelay)
+	client := h.Client
+	host1 := h.TestData.Host1
+	host2 := h.TestData.Host2
+
+	// start client and provider
+	require.NoError(t, h.Provider.Start(ctx))
+	require.NoError(t, client.Start(ctx))
+
+	// set ask price where we'll accept any price
+	err := h.Provider.SetAsk(big.NewInt(0), big.NewInt(0), 50_000)
+	require.NoError(t, err)
+
+	// wait for provider to enter deal transferring state and stop
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	_ = h.Provider.SubscribeToEvents(func(event storagemarket.ProviderEvent, deal storagemarket.MinerDeal) {
+		if event == storagemarket.ProviderEventDataTransferInitiated {
+			ev := storagemarket.ProviderEvents[event]
+			t.Logf("event %s has happened on provider, shutting down provider", ev)
+			require.NoError(t, h.TestData.MockNet.UnlinkPeers(host1.ID(), host2.ID()))
+			require.NoError(t, h.TestData.MockNet.DisconnectPeers(host1.ID(), host2.ID()))
+			require.NoError(t, h.Provider.Stop())
+			wg.Done()
+		}
+	})
+
+	result := h.ProposeStorageDeal(t, &storagemarket.DataRef{TransferType: storagemarket.TTGraphsync, Root: h.PayloadCid}, false, false)
+	proposalCid := result.ProposalCid
+	t.Log("storage deal proposed")
+
+	wg.Wait()
+	t.Log("provider has been shutdown the first time")
+
+	// Assert client state
+	cd, err := client.GetLocalDeal(ctx, proposalCid)
+	require.NoError(t, err)
+	t.Logf("client state after stopping is %s", storagemarket.DealStates[cd.State])
+	require.True(t, cd.State == storagemarket.StorageDealStartDataTransfer || cd.State == storagemarket.StorageDealTransferring)
+
+	// RESTART ONLY PROVIDER
+	h.createNewProvider(t, ctx, h.TestData, h.TempFilePath)
+	pds, err := h.Provider.ListLocalDeals()
+	require.NoError(t, err)
+	t.Logf("provider state after stopping is %s", storagemarket.DealStates[pds[0].State])
+	require.Equal(t, storagemarket.StorageDealTransferring, pds[0].State)
+
+	expireWg := sync.WaitGroup{}
+	expireWg.Add(1)
+	_ = h.Provider.SubscribeToEvents(func(event storagemarket.ProviderEvent, deal storagemarket.MinerDeal) {
+		if event == storagemarket.ProviderEventDealExpired {
+			expireWg.Done()
+		}
+	})
+
+	expireWg.Add(1)
+	_ = client.SubscribeToEvents(func(event storagemarket.ClientEvent, deal storagemarket.ClientDeal) {
+		if event == storagemarket.ClientEventDealExpired {
+			expireWg.Done()
+		}
+	})
+
+	// sleep so go-data-transfer gives up on retries after creating new connection
+	time.Sleep(15 * time.Second)
```

Comment:
is this waiting for go-graphsync or go-data-transfer to give up? I think go-graphsync, right?

that's fine. I should provide a way to disable retries in graphsync to prevent this.

anytime we're putting in a sleep longer than like a second in a test we may want put a check to skip in `--short` mode so you can run all tests locally with some expectation they'll run relatively fast.

```
   if testing.Short() {
        t.Skip("skipping test in short mode.")
    }
```